cpp/src/mmap.cpp

```diff
@@ -0,0 +1,443 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <sys/mman.h>
+
+#include <cstddef>
+#include <cstdlib>
+#include <future>
+#include <optional>
+#include <stdexcept>
+#include <type_traits>
+#include <unordered_map>
+
+#include <kvikio/error.hpp>
+#include <kvikio/mmap.hpp>
+#include <kvikio/nvtx.hpp>
+#include <kvikio/parallel_operation.hpp>
+#include <kvikio/posix_io.hpp>
+#include <kvikio/utils.hpp>
+#include "kvikio/file_utils.hpp"
+
+#include "utils.hpp"
+
+namespace kvikio {
+
+namespace detail {
+/**
+ * @brief Prevent the compiler from optimizing away the read of a byte from a given address
+ *
+ * @param addr The address to read from
+ */
+void disable_read_optimization(void* addr)
+{
+  auto addr_byte = static_cast<std::byte*>(addr);
+  std::byte tmp{};
+  asm volatile("" : "+r,m"(tmp = *addr_byte) : : "memory");
+}
+
+/**
+ * @brief Change an address `p` by a signed difference of `v`
+ *
+ * @tparam Integer Signed integer type
+ * @param p An address
+ * @param v Change of `p` in bytes
+ * @return A new address as a result of applying `v` on `p`
+ *
+ * @note It is UB in C++ when the initial pointer and the resulting pointer do not point to the
+ * elements from the same array. This UB is considered acceptable here due to lack of a better
+ * alternative.
+ */
+template <typename Integer>
+void* pointer_add(void* p, Integer v)
+{
+  static_assert(std::is_integral_v<Integer>);
+  return static_cast<std::byte*>(p) + v;
+}
```

Comment:
What about something like the following, I think that is implementation-defined, but not undefined?

```c++
template <typename Integer>
void* pointer_add(void* p, Integer v)
{
  static_assert(std::is_integral_v<Integer>);

  auto addr = reinterpret_cast<std::uintptr_t>(p);
  addr += static_cast<std::uintptr_t>(v);
  return reinterpret_cast<void*>(addr);
}
```

Reply:
> whether it constitutes a UB if the conversion is not a round-trip to a fixed object in the uintptr_t approach.

Ahh, you are right. 
The following is legal if the returned pointer remain within the bounds of the same allocated object (or one-past-the-end) as `p`:

```c++
void* pointer_add(void* p, std::ptrdiff_t v)
{
    return static_cast<byte*>(p) + v;
}
```